### DISCUSSION TITLE
[varLib.merge] designspace with sparse masters, and sparse features?

### DISCUSSION THREAD

**Discussion by justvanrossum** [fonttools/fonttools#2690] [General]
I am trying (and failing) to build a variable font that has some intermediate masters that are sparse (the only contain a subset of the glyphs), but do have features for that subset (kerning). But this set of features is a _subset_ of the features of the non-sparse masters. Is this a configuration that should work, or is this unsupported?

Given the error below, I am suspecting "unsupported", but I'm not certain, and definitely won't rule out I'm doing something wrong in my setup.

The sparse masters are in separate ufos.

The specific error I'm seeing is this (after applying #2689):

```
fontmake: Error: In 'MySans[opsz,wght].designspace': Generating fonts from Designspace failed: 

Couldn't merge the fonts, because some values were different, but should have
been the same. This happened while performing the following operation:
GSUB.table.FeatureList.FeatureCount

The problem is likely to be in My Sans JumpSmall:
Expected to see .FeatureCount==9, instead saw 2
```

**Comment by justvanrossum** [comment]
While fiddling with https://github.com/fonttools/fonttools/pull/2689#issuecomment-1181766476, I managed to get this error:
```
Incompatible features between masters.
Expected: aalt, fwid, locl, locl, locl, locl, pwid, vert, vrt2.
Got: aalt, vrt2.
```

**Comment by behdad** [comment]
This current is not supported. GSUB/GPOS of all masters that participate must have the same features and lookups. It becomes hard, but not impossible, to do otherwise.

**Comment by justvanrossum** [comment]
Thank you!

**Comment by justvanrossum** [comment]
Perhaps useful for others: I managed to work around this limitation using these steps:

- copy features.fea from default master into a separate .fea file
- delete all features.fea files from ALL masters (my ufos are temp files, be careful with your true sources)
- build VF without the .fea features (will still build automatic GPOS features such as kern, mark, mkmk, etc.)
- add the GSUB features to the build VF from the copied .fea file

This works because my project does not contain manually authored GPOS, and it doesn't do FeatureVariations.

All GSUB features are the same per master, so this should save some build time, too.

I needed an additional workaround for some feaLib.builder.Builder pickiness: it complains about not building GDEF, even when adding only GSUB to a VF. 
```python
import sys
from fontTools.feaLib.builder import Builder
from fontTools.ttLib import TTFont


def addOpenTypeFeatures(font, featurefile, tables=None, debug=False):
    builder = Builder(font, featurefile)
    builder.varstorebuilder = None  # Workaround for feaLib weirdness
    builder.build(tables=tables, debug=debug)


if __name__ == "__main__":
    fontfile, featurefile = sys.argv[1:]
    font = TTFont(fontfile)
    addOpenTypeFeatures(font, featurefile, tables=["GSUB"])
    font.save(fontfile)
```

Reason for this workaround:

https://github.com/fonttools/fonttools/blob/02a669ed6313d5479c0e6b176973cefa112cd32b/Lib/fontTools/feaLib/builder.py#L227-L234

Also made me find #2692
